event_profiling.h

```diff
@@ -0,0 +1,169 @@
+#ifndef USE_EVENT_PROFILING
+#define USE_EVENT_PROFILING 0
+#endif
+
+#ifndef DEBUG_EVENT_PROFILING
+#define DEBUG_EVENT_PROFILING 0
+#endif
+
+#ifndef RUBY_EVENT_PROFILING_H
+#define RUBY_EVENT_PROFILING_H 1
+
+#if USE_EVENT_PROFILING
+
+#include "ruby/internal/config.h"
+
+#include "vm_core.h"
+
+#include <pthread.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <time.h>
+#include <unistd.h>
+
+#if !defined(__GNUC__) && USE_EVENT_PROFILING
+#error "USE_EVENT_PROFILING is not supported by other than __GNUC__"
+#endif
+
+/* To avoid using realloc, allocate large heap memory ahead of execution. */
+typedef struct event_profiling_config
+{
+    int max_ractors;
+    int max_ractor_events;
+    int max_call_stack_depth;
+} event_profiling_config_t;
+
+typedef enum profiling_event_phase
+{
+    PROFILING_EVENT_PHASE_BEGIN,
+    PROFILING_EVENT_PHASE_END,
+    PROFILING_EVENT_PHASE_SNAPSHOT,
+} profiling_event_phase_t;
+
+typedef struct profiling_event
+{
+    /* Title: file:function(id) Args: line, ractor */
+    char *file;
+    char *function;
+    int   line;
+    int   ractor;
+    int   id;
+
+    pid_t pid;
+    pid_t tid;
+
+    profiling_event_phase_t phase;
+
+    char *snapshot_reason;
+
+    time_t timestamp;
+} profiling_event_t;
+
+/* Allocate an event list for each Ractor */
+typedef struct profiling_event_list
+{
+    int last_event_id;
+    int tail;
+
+    struct
+    {
+        int  top;
+        int *event_indexes;
+    } call_stack;
+
+    profiling_event_t *events;
+} profiling_event_list_t;
+
+typedef struct profiling_event_bucket
+{
+    profiling_event_list_t * system_init_event_list;
+    profiling_event_list_t **ractor_profiling_event_lists;
+    int                      ractors;
+} profiling_event_bucket_t;
+
+/* Global configuration */
+extern event_profiling_config_t *rb_event_profiling_config;
```

Comment:
Global symbols need to be prefixed to get rid of possible conflicts.

Reply:
I don't think we need to expose these global symbols for now so I just removed them from `event_profiling.h`

Reply:
I mean all symbols in this header file.

Reply:
I've got them done. Could you please check again? I think they should be prefixed with `rb_`.

Reply:
Looks fine, `static` symbols didn't need though.